Fixes #21737 - move about page to react

amirfefer opened the pull request:
About page uses 3 different old js files in order to check the status of compute resources and smart proxies.
I've extracted this status rendering to a react component `<About>`, in addition, I've united the UI style (by using Icons instead of labels).

### new components from **pf-react**:
- [x]  Empty State
- [x]  Tabs
- [x]  Table

###  ~~Todo~~
- [x]  tests
### updated gif (March) :
![After](https://user-images.githubusercontent.com/11807069/37245633-e105ed9c-24a3-11e8-9a07-14524fc20a3b.gif)

### before:
![nov 22 2017 6-10 pm](https://user-images.githubusercontent.com/11807069/33137426-962d0a00-cfb0-11e7-94e2-ccdb9d23930c.gif)

amirfefer committed:
Fixes #21737 - move about page to react

ohadlevy commented on app/views/about/index.html.erb:
```diff
@@ -16,105 +15,28 @@
-            </table>
-          <% end %>
+          <%= data = @smart_proxies.map {|proxy| {:id => {:name => proxy.name, :id => proxy.id},
+                                                  :features => h(proxy.features.map(&:name).to_sentence)}} %>
```
```h``` is probably no longer required (it's the default).

ohadlevy commented on app/views/about/index.html.erb:
```diff
@@ -16,105 +15,28 @@
-                </tbody>
-              </table>
-            <% end %>
+            <%= data = @compute_resources.map {|compute| {:id => {:name => compute.name, :id => compute.id},
```
when there are none, you get a generic message ````No data found``` - we should use the PF empty state instead.

ohadlevy commented on config/routes.rb:
```diff
@@ -404,7 +404,7 @@
           post 'cluster_selected'
           get 'resource_pools'
-          post 'ping'
+          get 'ping'
```
doesnt this break CR test connection?

ohadlevy commented on webpack/assets/javascripts/react_app/common/helpers.js:
```diff
@@ -6,4 +6,12 @@ export default {
+    return `/${controller}/${id ? `${id}/` : ''}${action}`;
+  },
+  getQueryParam(query) {
+    const urlParams = new URLSearchParams(window.location.search);
```
how is this different from the urijs lib that we currently use? - also does this work across browsers?

amirfefer replied:
It becomes a standard interface, most of the browsers support it, and there's a [polyfill](https://github.com/WebReflection/url-search-params)

ohadlevy replied:
should we replace than the urijs?

ohadlevy commented on webpack/assets/javascripts/react_app/components/about/compute/index.js:
```diff
@@ -0,0 +1,45 @@
+  render() {
+    const columns = [
+      {
+        Header: 'Name',
```
please extract all strings

ohadlevy commented on webpack/assets/javascripts/react_app/components/about/compute/index.js:
```diff
@@ -0,0 +1,45 @@
+        Header: 'Name',
+        accessor: 'id',
+        Cell: props => (
+          <a href={helpers.urlBuilder('compute_resources', '', props.value.id)}>
```
you can import urlBruilder directly, e.g.
```js
import { urlBuilder } from '../helpers';
```

ohadlevy commented on webpack/assets/javascripts/react_app/components/about/provider/index.js:
```diff
@@ -0,0 +1,24 @@
+import React from 'react';
+import Table from '../../common/table';
+import Icon from '../../common/Icon';
```
PF-react now has icon component - maybe we should consider using that instead and remove our own?

ohadlevy commented on webpack/assets/javascripts/react_app/components/common/table/index.js:
```diff
@@ -0,0 +1,54 @@
+import 'react-table/react-table.css';
+import helpers from '../../../common/helpers';
+
+const SETTING_PER_PAGE = 10; // TODO: use foreman per_page settings
```
current UI default is 20

ohadlevy commented on webpack/assets/javascripts/react_app/components/common/table/index.js:
```diff
@@ -0,0 +1,54 @@
+import React from 'react';
+import ReactTable from 'react-table';
+import 'react-table/react-table.css';
```
table style doesn't look like patternfly...

ohadlevy commented on webpack/assets/javascripts/react_app/components/common/table/index.js:
```diff
@@ -0,0 +1,54 @@
+        className="-striped"
+        defaultPageSize={
+          showPagination
+            ? Number(helpers.getQueryParam('per_page')) || Math.min(rows, SETTING_PER_PAGE)
```
why deal with pagination in this PR if there is no code that uses it yet?

ohadlevy commented on webpack/assets/javascripts/react_app/components/common/table/index.js:
```diff
@@ -0,0 +1,54 @@
+        data={data}
+        columns={columns}
+        // i18n
+        previousText={__('Previous')}
```
if you are adding pagination support - it should match pf

ohadlevy commented on webpack/assets/javascripts/react_app/components/common/table/index.js:
```diff
@@ -0,0 +1,54 @@
+        pageText={__('page')}
+        ofText={__('of')}
+        rowsText={__('rows')}
+        noDataText={__('No data found')}
```
should probably use empty state optionally (in case the table is the main part of the view).

ohadlevy commented on webpack/assets/javascripts/react_app/components/common/table/index.js:
```diff
@@ -0,0 +1,54 @@
+                ? rowData
+                : undefined,
+            style: {
+              textAlign: 'center',
```
why pass style here?

ohadlevy commented on webpack/assets/javascripts/react_app/components/componentRegistry.js:
```diff
@@ -8,6 +8,10 @@ import PowerStatus from './hosts/powerStatus/';
 import NotificationContainer from './notifications/';
 import ToastsList from './toastNotifications/';
 import StorageContainer from './hosts/storage/vmware/';
+import AboutComputeTable from './about/compute';
```
TBH: I am not too happy about having so many entry level components, is there any chance we can take over multiple tabs in one hoc ? this will allow to have all of the "logic" in one place vs across multiple components

ohadlevy commented on webpack/assets/javascripts/react_app/redux/actions/status/index.js:
```diff
@@ -0,0 +1,12 @@
+import { STATUS_REQUEST, STATUS_SUCCESS, STATUS_FAILURE } from '../../consts';
```
shouldnt these const be more specific? e.g. HOST_STATUS or COMPUTE_RESOURCE_STATUS?

amirfefer replied:
I've extracted it out of the tree, something like this:

STATUS
    -COMPUTE_RESOURCE
        [-X...] 
    -SMART_PROXY
        [-Y...]

ohadlevy commented on webpack/assets/javascripts/react_app/redux/reducers/status/index.js:
```diff
@@ -0,0 +1,19 @@
+
+  switch (action.type) {
+    case STATUS_REQUEST:
+      return state;
```
why? shouldn't this set the state so that the request has started?

ohadlevy commented on webpack/assets/javascripts/react_app/redux/reducers/status/index.js:
```diff
@@ -0,0 +1,19 @@
+    case STATUS_REQUEST:
+      return state;
+    case STATUS_SUCCESS:
+      return state.set(payload.id, { ...payload });
```
I'm a bit worried that ID's can override each other, while I've seen you've prefixed it, the fact is that these actions are used for multiple objects, can confuse a future developer?

ohadlevy reviewed (changes requested):
nice @amirfefer - first review round..

ohadlevy commented:
@amirfefer also note the lack of tests + current failing tests.

sharvit commented:
@amirfefer It's is great to see more and more code moving to react.
I would love to see the whole `about/index.html.erb` moving to react, including the tabs and the layout.
Then, you will need to add only 1 component to the registry (instead of 4).

You can use `react-bootstrap` for the tabs:
https://react-bootstrap.github.io/components.html#tabs

The row and col can be replaced with:
https://react-bootstrap.github.io/components.html#grid

Those components will be apart of `patternfly-react` in the future so it will be easy to move them later, probably by only change the import statement.

ohadlevy commented on app/views/about/index.html.erb:
```diff
@@ -16,105 +15,28 @@
-              <% @smart_proxies.each do |proxy| %>
-                <tr class="proxy-show"  data-url="<%= ping_smart_proxy_path(proxy) %>">
-                  <td><%= link_to_if_authorized proxy.name, hash_for_smart_proxy_path(proxy) %></td>
-                  <td><%=h proxy.features.map(&:name).to_sentence %></td>
```
I wonder how translation worked here before?

amirfefer commented:
@ohadlevy 
I've changed the table library to [react-bootstrap-table](http://allenfang.github.io/react-bootstrap-table/) for an easy integration with pf styling. (`react-table` doesn't use the regular table, th, td tr tags)
I've created a new Tabs component, so there's only one registered component (instead of 4)
I've implemented a new `empty state` component that renders when no data instead of a `no data` string.

empty state component:
![empty_state](https://user-images.githubusercontent.com/11807069/33336319-286a35d4-d478-11e7-88a5-563743fa2dcb.png)

bootsrap & pf table style:
![new_table_style](https://user-images.githubusercontent.com/11807069/33336321-2c178d08-d478-11e7-875e-b2f9df6adcc8.png)

ohadlevy commented:
@amirfefer i think you forgot to push your new implementation.

regardless, I would suggest to send PR's to patternfly-react repo with all of pf related components (tabs, emptystate etc)

ohadlevy commented on app/assets/stylesheets/base.scss:
```diff
@@ -250,7 +250,7 @@ select {
 table {
   clear: both;
-  margin-bottom: 6px !important;
+  //margin-bottom: 6px !important;
```
?

ohadlevy commented on app/assets/stylesheets/base.scss:
```diff
@@ -335,7 +335,7 @@ table {
 
 #pagination {
-  margin-top: -6px;
+  margin-top: -19px !important;
```
why change style for all pagination?

ohadlevy commented on app/helpers/about_helper.rb:
```diff
@@ -0,0 +1,31 @@
+module AboutHelper
+  def plugins
```
if you are creating a controller per action, then I would expect AboutPluginController with only index action.

having said that, I would default to use the API first and only if it doesnt fit create a UI end point

amirfefer replied:
I've created About controller namespace, inspired by [this](http://jeromedalbert.com/how-dhh-organizes-his-rails-controllers/) post.

ohadlevy replied:
in that case why do you need the helper than?
also.. can't we simply use the existing api?

tstrachota replied:
Using an api endpoint would definitely be more systemic. Such endpoint doesn't exist at the moment. We can only list proxies and compute profiles separately. How about moving it to a new api endpoint? It would help with http://projects.theforeman.org/issues/3036.

But I understand it's adds some work and there are probably issues with higher priority. So if you don't have space for moving it to API, I can live with the current state ;)

ohadlevy commented on app/views/about/index.html.erb:
```diff
@@ -1,122 +1,13 @@
-        </div>
-      </div>
+      <div id="tabs"></div>   
+      <%= mount_react_component('About', '#tabs', {:compute => compute_resources, :proxy => proxies,
```
probably makes sense to combine it under a method, so in theory we can expand it later vs hard code it. e.g.

```ruby
def about_data
  {:compute => compute_resources, :proxy => proxies,...
end
```

ohadlevy commented on package.json:
```diff
@@ -68,12 +68,13 @@
     "multiselect": "~0.9.12",
     "patternfly": "^3.29.5",
-    "patternfly-react": "^0.8.1",
+    "patternfly-react": "^0.11.0",
```
all package upgrades seems not related?

amirfefer replied:
I've upgraded in order to use Icon component

ohadlevy commented on webpack/assets/javascripts/react_app/components/about/compute/index.js:
```diff
@@ -0,0 +1,50 @@
+  render() {
+    const columns = [
+      {
+        header: 'Name',
```
string not extracted?

amirfefer replied:
extracted already in `<Table>` component

ohadlevy commented on webpack/assets/javascripts/react_app/components/about/index.js:
```diff
@@ -0,0 +1,25 @@
+    } = this.props.data;
+    const tabs = [__('Smart Proxies'), __('Available Providers'), __('Compute Resources'), __('Plugins')];
+
+    return <ForemanTabs id='about_tabs' tabs={tabs} >
```
this looks really close to how react-router works.. I wonder if we should combine the two? (so if you refresh you come back to the same tab)

amirfefer replied:
sounds great, though this PR becomes too large, maybe should we combine react-router in a different PR?

ohadlevy commented on webpack/assets/javascripts/react_app/components/common/emptyState/index.js:
```diff
@@ -0,0 +1,63 @@
+        </p>
+        <div className="blank-slate-pf-main-action">
+          <button
+            onClick={e => window.Turbolinks.visit(action.url)}
```
wouldnt turbolink catch these regardless?

ohadlevy commented on webpack/assets/javascripts/react_app/components/common/emptyState/index.js:
```diff
@@ -0,0 +1,63 @@
+        <h1>{header}</h1>
+        <p>{description}</p>
+        <p>
+          Learn more about this{' '}
```
not extracted..

ohadlevy reviewed (changes requested):
I ran out of time, but here is some feedback

ohadlevy commented on app/views/about/plugins_empty_state/index.json.rabl:
```diff
@@ -0,0 +1,6 @@
+object false
+node do |u|
+  { :header => N_("Plugin"), :description => N_("Plugins are tools to extend and modify the functionality of Foreman. Plugins offer custom functions and features so that each user can tailor their environment to their specific needs."),
```
this makes me wonder why not simply store this inside a component ? e.g. whats the value of making an api call to get static values?

ohadlevy commented on package.json:
```diff
@@ -72,6 +72,7 @@
     "raf": "^3.4.0",
     "react": "^16.0.0",
     "react-bootstrap": "^0.31.0",
+    "react-bootstrap-table": "^4.1.5",
```
I'm in mixed feelings about this library - didn't  you prefer the other one? :)

amirfefer replied:
The other library doesn't use bootstrap table classes, and more then that it doesn't use tables tags at all , so it would be difficult to add bootstrap and patternfly style.

ohadlevy commented on webpack/assets/javascripts/react_app/components/about/compute/index.js:
```diff
@@ -0,0 +1,50 @@
+import Table from '../../common/table';
+import helpers from '../../../common/helpers';
+
+class AboutComputeTable extends React.Component {
```
nitpick: please call it compute resource (no point in shorting the name imho)

amirfefer commented:
@ohadlevy , @sharvit , @waldenraines 

I updated this PR to use these **patternfly-react** components:
- Table
- Empty state
- Tabs

![about - edited 1](https://user-images.githubusercontent.com/11807069/37245633-e105ed9c-24a3-11e8-9a07-14524fc20a3b.gif)

amirfefer commented:
### Performance:
post react:
![screenshot from 2018-03-10 20-25-27](https://user-images.githubusercontent.com/11807069/37245519-f3967c1c-24a1-11e8-9d5f-8b44adb5c33f.png)
pre react:
![screenshot from 2018-03-10 20-29-28](https://user-images.githubusercontent.com/11807069/37245522-fa6681ae-24a1-11e8-9e25-3032c2470e83.png)

amirfefer commented:
I've added couple of tests.

amirfefer commented on test/integration/about_test.rb:
```diff
@@ -14,31 +14,17 @@ class AboutIntegrationTest < IntegrationTestWithJavascript
-    assert page.has_content?("Version")
-  end
-
-  test "about page proxies should have version" do
```
already tested [here](https://github.com/theforeman/foreman/pull/5024/files#diff-15b6dd3ae736678ae8923d366096fea5R49)

amirfefer commented on test/integration/about_test.rb:
```diff
@@ -14,31 +14,17 @@ class AboutIntegrationTest < IntegrationTestWithJavascript
 
   test "about page" do
     assert_index_page(about_index_path,"About", nil, false, false)
-    wait_for_ajax
```
`axios` breaks `wait_for_ajax` (We are using axios instead of jquery in react components)
it's redundant here.

tstrachota commented on webpack/assets/javascripts/react_app/components/about/plugin/schema.js:
```diff
@@ -0,0 +1,44 @@
+      formatters: [headerFormat],
+    },
+    cell: {
+      formatters: [cellFormat],
```
Plugin names used to be links. Did you remove it by intention?

waldenraines commented on webpack/assets/javascripts/react_app/components/common/table/index.js:
```diff
@@ -0,0 +1,40 @@
+  }
+
+  render() {
+    const { columns, data } = this.props;
```
Thoughts on renaming `data` to `rows`?  I think this would make more sense since that's what these are.

waldenraines commented on webpack/assets/javascripts/react_app/components/common/emptyState/index.js:
```diff
@@ -0,0 +1,70 @@
+        {secondayActions && (
+          <PfEmptyState.Action secondary>
+            {secondayActions.map(item => (
+              <Button href={action.url} title="Perform an action">
```
This string isn't translated.

tstrachota replied:
I guess that "Perform an action" is just copy leftover anyway.

waldenraines commented on webpack/assets/javascripts/react_app/components/about/plugin/index.js:
```diff
@@ -0,0 +1,20 @@
+  const emptyStateData = {
+    header: __('Plugin'),
+    description: __('Plugins are tools to extend and modify the functionality of Foreman. Plugins offer custom functions and features so that each user can tailor their environment to their specific needs.'),
+    docUrl: 'https://www.theforeman.org/manuals/1.15/index.html#Plugins',
```
Don't we also need a documentation link text here?  I think we want the ability to customize the text of this link.

amirfefer replied:
good point, I added prop `documentationText` with a default value of Documentation

waldenraines commented on webpack/assets/javascripts/react_app/components/about/plugin/index.js:
```diff
@@ -0,0 +1,20 @@
+    header: __('Plugin'),
+    description: __('Plugins are tools to extend and modify the functionality of Foreman. Plugins offer custom functions and features so that each user can tailor their environment to their specific needs.'),
+    docUrl: 'https://www.theforeman.org/manuals/1.15/index.html#Plugins',
+    action: { title: __('Get a plugin'), url: 'https://projects.theforeman.org/projects/foreman/wiki/List_of_Plugins' },
```
Using url here will cause a full page reload instead of a transition if using `react-router`.  How do you feel about adding a `to` property here for use with `react-router`?  Obviously we would also need to include react-router into core so I'm fine if we tackle this outside of this PR.

waldenraines replied:
Or perhaps accept components for the actions instead?  That would make this more flexible at the cost of some repetition.

ohadlevy replied:
we can also use Turbolinks.visit(url) as we do in many other places.

waldenraines commented on webpack/assets/javascripts/react_app/components/about/proxies/schema.js:
```diff
@@ -0,0 +1,78 @@
+  {
+    property: 'id',
+    header: {
+      label: 'Status',
```
Should these labels be translated?

waldenraines commented on webpack/assets/javascripts/react_app/components/about/compute/schema.js:
```diff
@@ -0,0 +1,57 @@
+import { headerFormat, cellFormat } from '../../common/table';
+import helpers from '../../../common/helpers';
+
+export const columns = [
```
Any thoughts to how this will work with selectable columns?

amirfefer replied:
It should work,
we can follow patternfly-react [example](https://github.com/patternfly/patternfly-react/blob/master/src/components/Table/__mocks__/mockClientPaginationTable.js)

waldenraines reviewed (changes requested):
Used the table/empty state wrapper and schema in my work on https://github.com/Katello/katello/pull/7230.  

I think I like the approach and just have a few comments.

tstrachota commented on webpack/assets/javascripts/react_app/components/common/tabs/index.js:
```diff
@@ -0,0 +1,33 @@
+        <Nav bsClass="nav nav-tabs nav-tabs-pf" justified>
+          {tabs.map((tab, index) => (
+            <NavItem key={index} eventKey={index}>
+              {__(tab)}
```
This is translated already in https://github.com/theforeman/foreman/pull/5024/files#diff-26f92dbe854c0b8737e453ea192ebe2fR12
Translations should be handled outside of components if possible, so remove this one, please.

tstrachota commented on webpack/assets/javascripts/react_app/components/common/emptyState/index.js:
```diff
@@ -0,0 +1,70 @@
+    if (data) {
+      this.setState({ data });
+    } else {
+      API.get(url).then(
```
Why do we need to fetch the data from server? Does it contain anything else than translated strings?
Since a rabl template needs to be created anyway, can't we just drop the server side and create a wrapper empty state component for each model and save one request? It would also help with keeping the code at one place.

tstrachota commented on webpack/assets/javascripts/react_app/components/common/emptyState/index.js:
```diff
@@ -0,0 +1,70 @@
+        <PfEmptyState.Title>{header}</PfEmptyState.Title>
+        <PfEmptyState.Info>{__(description)}</PfEmptyState.Info>
+        <PfEmptyState.Help>
+          For more information please see{' '}
```
This needs to be translated

tstrachota commented on webpack/assets/javascripts/react_app/components/common/emptyState/index.js:
```diff
@@ -0,0 +1,70 @@
+        <PfEmptyState.Info>{__(description)}</PfEmptyState.Info>
+        <PfEmptyState.Help>
+          For more information please see{' '}
+          <a href={docUrl}>{__('documentation')}</a>
```
I'm afraid this is going to cause issues with translating languages with different word order. It effectively prevents from putting the word "documentation" anywhere else in the sentence.

ohadlevy replied:
there is also https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/common/DocumentationLink/index.js maybe we can combine it somehow?

tstrachota commented on webpack/assets/javascripts/react_app/components/common/emptyState/index.js:
```diff
@@ -0,0 +1,70 @@
+      <PfEmptyState>
+        <PfEmptyState.Icon type="pf" name={icon || 'add-circle-o'} />
+        <PfEmptyState.Title>{header}</PfEmptyState.Title>
+        <PfEmptyState.Info>{__(description)}</PfEmptyState.Info>
```
This should be translated outside of the component.

tstrachota commented on app/controllers/application_controller.rb:
```diff
@@ -36,6 +36,12 @@ def welcome
     not_found
   end
 
+  def empty_state
```
Loading the empty state setup from the server side seems unnecessarily complicated to me. See my comment in the empty state component please.

tstrachota reviewed (changes requested):
Thanks @amirfefer. It looks good visually and I like the status components.

Table labels need to be tranlsated. There's also couple of translation issues in the empty state component.

What benefits does loading the empty state settings from the server bring?

ohadlevy commented on webpack/assets/javascripts/react_app/components/about/compute/index.js:
```diff
@@ -0,0 +1,13 @@
+import { columns } from './schema';
+
+const AboutComputeTable = props => (
+      <Table
```
inconsistent spacing?

waldenraines commented:
@amirfefer updates on this?  Would like to get this in soon since it contains change that I am using in https://github.com/Katello/katello/pull/7230

amirfefer commented:
@waldenraines : ready for another review :)

amirfefer commented:
@tstrachota, thanks for your review.
I tend to agree with the empty state API endpoint, it's a bit redundant.

tstrachota commented on webpack/assets/javascripts/react_app/components/common/emptyState/index.js:
```diff
@@ -0,0 +1,43 @@
+    action,
+    secondayActions,
+  } = props;
+  const defaultDocumantion = __(`${documentationLabel} <a href=${docUrl}>${documentationButton}</a>`);
```
This is not going to work correctly. The string gets interpreted first so you'll get a string `"For more information..."` but translation files will contain key `"${documentationLabel} <a href=${docUrl}>${documentationButton}</a>"` and that's not going to match.

We need to use `sprintf` (for instance like here https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/common/reduxFormI18n.js#L12)

waldenraines replied:
Nitpick, spelling is incorrect, should be documentation not documantion.

tstrachota commented on webpack/assets/javascripts/react_app/components/common/emptyState/index.js:
```diff
@@ -0,0 +1,43 @@
+    description,
+    customDocumentation,
+    documentationLabel = 'For more information please see',
+    documentationButton = 'Documentation',
```
Are these defaults used anywhere? If yes, they should be translated as well.

tstrachota reviewed (changes requested):
Two more translation-related comments. Apart from that it works nicely!

waldenraines reviewed (approved):
Looks good to me.  Everything I tested worked and I also reused the table and empty state components successfully. 

@amirfefer please fix the build and hopefully we can get this merged soon.

tstrachota commented:
IMHO there are still 2 translation issues that deserve fixing - see my previous comments.

waldenraines commented:
> IMHO there are still 2 translation issues that deserve fixing - see my previous comments.

Yeah, for sure, I didn't mean this was approved without those changes but including them.

waldenraines commented on webpack/assets/javascripts/react_app/components/common/emptyState/index.js:
```diff
@@ -0,0 +1,43 @@
+    customDocumentation,
+    documentationLabel = 'For more information please see',
+    documentationButton = 'Documentation',
+    docUrl,
```
I still would prefer to see this as an object like the action.  We should be able to customize the text of this button instead of it always being "Documentation".

Something like this:

```javascript
documentation: {
  title: __('Learn more about adding Subscriptions to Allocations'), 
  url: 'http://redhat.com' 
},
```

waldenraines commented on webpack/assets/javascripts/react_app/components/common/table/index.js:
```diff
@@ -0,0 +1,40 @@
+          rows={rows}
+          rowKey={({ rowData, rowIndex }) => rowIndex}
+        />
+      </PfTable.PfProvider>
```
Do we want to include pagination with this?  Right now if you have pagination on your table and trigger the empty state you get this:

<img width="1217" alt="screen shot 2018-03-27 at 11 38 50 am" src="https://user-images.githubusercontent.com/4116405/37978179-73707474-31b3-11e8-8ed6-a92f3c44de64.png">

Was thinking maybe you could pass the pagination component into this directly so that way you maintain control of it from within your component.  In other words:

```javascript
const paginationRow = (
  <PaginationRow
    viewType="table"
    itemCount={subscriptions.pagination.total}
    pagination={subscriptions.pagination}
    onChange={onPaginationChange}
  />
);

<Table
  rows={subscriptions.results}
  columns={columns}
  emptyState={emptyStateData()}
  paginationRow={paginationRow}
/>
```

Then all this component would do is include the pagination within the ternary operator.

waldenraines replied:
I'm also okay with adding this in a future PR.

tstrachota replied:
I'm leaning towards fixing it in a future PR. It's not a blocker for this one that has been open for a long time already.

I'd propose a bit different solution. The empty state design is quite constant while the table implementation can vary a lot (inline edits, provider setup, etc.). I think that it would be more flexible to refactor (and maybe rename) the component so that it accepts children with default being a table. But as I've said, I prefer to solve it outside of this PR.

waldenraines replied:
> I'd propose a bit different solution. The empty state design is quite constant while the table implementation can vary a lot (inline edits, provider setup, etc.). I think that it would be more flexible to refactor (and maybe rename) the component so that it accepts children with default being a table. But as I've said, I prefer to solve it outside of this PR.

I also prefer this solution and am okay updating katello once we have it done.

ohadlevy commented:
[test foreman]

mmoll commented:
This needs at least a rebase :point_up:
Also, have a look into the failing travis test.

tstrachota commented:
@amirfefer ping, could you rebase, please?

xprazak2 commented:
@amirfefer, could you rebase please?

tstrachota commented:
@amirfefer ping

amirfefer commented:
It was among the latter in the priority tasks list, 
I'll rebase and  have a look on it soon.

amirfefer commented:
@tstrachota , @waldenraines 
Rebased, and ready for a another look :)

boaz0 commented on webpack/assets/javascripts/react_app/components/common/table/index.js:
```diff
@@ -0,0 +1,40 @@
+    return this.isEmpty() ? (
+      <EmptyState {...emptyState} />
+    ) : (
+      <PfTable.PfProvider
```
Hi @amirfefer, this looks good but what if I want to use `components` in `<PfTable.PfProvider>` - I am not sure I can do that through `Table`. In `Hardware Models` I am using it to set how the heading and cells will look like.

tstrachota replied:
This has been solved in Katello: https://github.com/Katello/katello/blob/master/webpack/move_to_foreman/components/common/table/components/Table.js That component is based on @amirfefer 's solution and is going to be moved to foreman soon. AFAIR the interfaces should be compatible.
So, if you don't mind @ripcurld, I wouldn't block merging this PR by that. It contains lot of other good stuff, has been open for quite some time and is quite ready to be merged.

boaz0 replied:
@tstrachota I didn't mean to block :smiley:  but just to raise a concern I have in [`Hardware Models`](https://github.com/theforeman/foreman/pull/5755).

tstrachota commented:
There's minor conflict again. I wasn't fast enough with the final review:( I'm sorry for that. Please rebase @amirfefer

amirfefer commented:
@tstrachota  - rebased :)

tstrachota commented on webpack/assets/javascripts/react_app/common/EmptyStates.js:
```diff
@@ -0,0 +1,38 @@
+  description: __('Foreman supports creating and managing hosts on a number of virtualization and cloud services - referred to as “compute resources” - as well as bare metal hosts.'),
+  documentation: {
+    // eslint-disable-next-line no-undef
+    url: `https://www.theforeman.org/manuals/${VERSION}/index.html#5.2ComputeResources`,
```
I found out that the docs link (and the other too) leads to 404. `VERSION` is `1.19.0` but the manual url has to contain only `1.19`. We probably need both `VERSION` and `VERSION_MAJOR` or something similar, maybe an object containing multiple version formats and with `toString()` defined.

tstrachota commented on webpack/assets/javascripts/react_app/components/common/emptyState/index.js:
```diff
@@ -0,0 +1,46 @@
+      {url &&
+        <PfEmptyState.Help>
+         {label}
+          <a href={url}> {buttonLabel} </a>
```
Nitpick:
this indend doesn't put any space between the text and the link. Spaces in `<a>` work but they look weird on hover:
![screenshot from 2018-07-09 15-57-09](https://user-images.githubusercontent.com/1186631/42454858-ef55cf10-8390-11e8-9f15-769dfae3dfa1.png)

Please indent correctly and remove spaces form the anchor. `{" "}` should help with preserving the space in between.

tstrachota reviewed (changes requested):
I found two more issues on the empty states. Hopefully they'll be easy to fix.
The rest looks good and worked fine for me.

mmoll commented:
@amirfefer could you rebase and fix the last two small comments from @tstrachota? :)

mmoll commented:
what's the status here?

tstrachota commented:
This needs a rebase again.

waldenraines commented:
@amirfefer any updates on this?

amirfefer commented:
@waldenraines  - basically i'd like to use table from this [PR](https://github.com/theforeman/foreman/pull/5755), will re-prioritize this soon :)

mmoll commented:
@amirfefer what's the status here?

mmoll commented:
any news here? :)

ares commented:
@amirfefer this needs a bigger rebase now, could you please let us know if you plan to look at this soon? If there's no activity before next triage (2 weeks), we'd lean towards closing this. It would be unfrotunate as there were just 2 small things to finish this, but of course it can be reopened later if your availability for this PR changes.

ares commented:
Unfortunatelly there was no reply for 2 weeks so we are closing this PR now. Please reopen if you're still interested in getting this in.
